### PR TITLE
Introduce platform.PostApplyHook interface and implement it for AKS cluster

### LINKED ISSUE #855: AKS: prometheus-operator gets installed before default storage class object is created
This causes the CI to fail, as `prometheus-operator` component never converges.

This is related to #559, however, to solve this one, we should perhaps make sure, that before AKS installation converges, default storage class becomes available. We could for example install the storage class on AKS explicitly via chart, to avoid waiting for the AKS to converge.

### PR DESCRIPTION
**Pull request by invidian**
See commit messages for the details.

Closes #855

### CODE REVIEW

File: pkg/platform/platform.go [hunk 1]
```diff
@@ -74,6 +74,13 @@ type Platform interface {
 	Meta() Meta
```
Comment (knrt10) [inline pkg/platform/platform.go:74]: Some nits on the commit message. You used `interface` twice at the end of line 3. Also comma on line 3. I think it should be `period`.

File: pkg/platform/aks/aks.go [hunk 2]
```diff
@@ -309,6 +315,48 @@ func (c *config) Initialize(ex *terraform.Executor) error {
...
+// waitForDefaultStorageClass wait until default storage class appears on a given cluster.
+// If it doesn't within defined time range, error is returned.
+func waitForDefaultStorageClass(sci v1.StorageClassInterface) error {
+	// AKS still uses annotation with .beta.
+	defaultStorageClassAnnotation := "storageclass.beta.kubernetes.io/is-default-class"
+
+	if err := wait.PollImmediate(retryInterval, timeout, func() (done bool, err error) {
+		scs, err := sci.List(context.TODO(), metav1.ListOptions{})
+		if err != nil {
+			return false, fmt.Errorf("getting storage classes %w", err)
```
Comment (knrt10) [inline pkg/platform/aks/aks.go:343]:
```suggestion
			return false, fmt.Errorf("getting storage classes: %w", err)
```

**Review by knrt10 — COMMENTED** [review]
Just a NIT, rest LGTM.

File: pkg/platform/aks/aks.go (hunk 2 above)
Comment (invidian) [reply pkg/platform/aks/aks.go]: Oops, fixed thanks :)

File: pkg/platform/platform.go (hunk 1 above)
Comment (invidian) [reply pkg/platform/platform.go:74]: I _think_ I fixed it :smile: Please have a look again.

**Review by knrt10 — DISMISSED** [review]
+1

File: pkg/platform/aks/aks.go [hunk 3]
```diff
@@ -309,6 +315,48 @@ func (c *config) Initialize(ex *terraform.Executor) error {
...
+	client, err := k8sutil.NewClientset(kubeconfig)
+	if err != nil {
+		return fmt.Errorf("creating clientset from kubeconfig: %w", err)
+	}
+
+	return waitForDefaultStorageClass(client.StorageV1().StorageClasses())
+}
+
+// waitForDefaultStorageClass wait until default storage class appears on a given cluster.
+// If it doesn't within defined time range, error is returned.
```
Comment (iaguis) [inline pkg/platform/aks/aks.go:335]:
```suggestion
// waitForDefaultStorageClass waits until the default storage class appears on a given cluster.
// If it doesn't appear within a defined time range, an error is returned.
```

File: pkg/platform/platform.go [hunk 4]
```diff
@@ -74,6 +74,13 @@ type Platform interface {
...
 }
 
+// PostApplyHook is an interface, which platforms may optionally implement. If they do so,
+// after Terraform finishes applying, kubeconfig content will be passed to this function and
+// executed. This allows running sanity checks on the newly created cluster.
```
Comment (iaguis) [inline pkg/platform/platform.go:79]: I'd simplify this a bit.

```suggestion
// PostApplyHook runs code after Terraform finishes applying. This allows
// running sanity checks on the newly created cluster. Implementing this
// interface is optional for platforms.
```

**Review by iaguis — COMMENTED** [review]
Looks OK but I'd like to have @johananl's review before merging.

**Review by knrt10 — APPROVED** [review]

**Comment by johananl** [comment]
I'm fine with this implementation for now if it solves the problem for us. Once we refactor the platform interface I think it could make more sense to have a method called `PostApplyHook()` (or even `PostApplyHooks()`?) **on the platform interface** which optionally returns the hook(s) to be executed. Doing so would make things more explicit compared to implementing another interface.

**Review by iaguis — APPROVED** [review]

**Comment by invidian** [comment]
>I'm fine with this implementation for now if it solves the problem for us. Once we refactor the platform interface I think it could make more sense to have a method called PostApplyHook() (or even PostApplyHooks()?) on the platform interface which optionally returns the hook(s) to be executed. Doing so would make things more explicit compared to implementing another interface.

Since the hook is optional, I think it's nice to have it separate, as it allows us to avoid adding boilerplate code to all platforms. This is not something every platform MUST implement, like e.g. generating Terraform code or Terraform plan (this could actually be optional too).

**Comment by johananl** [comment]
> > I'm fine with this implementation for now if it solves the problem for us. Once we refactor the platform interface I think it could make more sense to have a method called PostApplyHook() (or even PostApplyHooks()?) on the platform interface which optionally returns the hook(s) to be executed. Doing so would make things more explicit compared to implementing another interface.
> 
> Since the hook is optional, I think it's nice to have it separate, as it allows us to avoid adding boilerplate code to all platforms. This is not something every platform MUST implement, like e.g. generating Terraform code or Terraform plan (this could actually be optional too).

The price you're paying is clarity: having an interface which can be optionally implemented leaves things implicit and more "magical" as far as the caller is concerned. It also doesn't make sense to me conceptually at the abstraction level: A platform is a type of a hook? :thinking: This is very different from saying "AWS is a type of platform".

I have no problem returning `nil` or an empty slice when no hooks are necessary on all platforms. To me this is much better than bending the abstraction to avoid "duplicating" a single line per platform.

**Comment by invidian** [comment]
>I have no problem returning nil or an empty slice when no hooks are necessary on all platforms. To me this is much better than bending the abstraction to avoid "duplicating" a single line per platform.

Well, that's exectly the boilerplate I'm talking about. And it's not single line, but at least 4 :smile: 

>The price you're paying is clarity: Having an interface which can be optionally implemented leaves things implicit and more "magical" as far as the caller is concerned. It also doesn't make sense to me conceptually at the abstraction level: "A platform is a type of hook?" thinking This is very different from saying "AWS is a type of platform". Go interfaces are built for certain purposes and I don't think this use case is one of them. I think a hint which supports this is that you had to type-assert the interface rather than just executing a generic method.

Yeah, I agree that abstraction wise this is not great. (Actually, initially my interface was named `PlatformPostApplyHook`, but then linter complains, that you do `platform.PlatformPostApplyHook`.

And what do you think about:
```go
type PlatformWithPostApplyHook interface {
  Platform
  PostApplyHook(kubeconfig []byte) error
}
```

I think it provides much better context then, as it extends the `Platform` interface and optionally ships this extra method. Name is something we should discuss of course :)

**Comment by iaguis** [comment]
I think this discussion doesn't belong to this PR :)

**Comment by invidian** [comment]
>I think this discussion doesn't belong to this PR :)

Well, I consider now changing this newly introduced interface to the version from https://github.com/kinvolk/lokomotive/pull/886#issuecomment-684885334 :smile:

**Comment by johananl** [comment]
For the sake of moving this forward @invidian, please do whatever makes sense to you which solves the problem *for now*, and I'll do my best to integrate whatever you come up with into the new platform interface in a way which makes sense. We can continue the discussion there.

**Review by knrt10 — APPROVED** [review]

**Review by iaguis — APPROVED** [review]

**Pull request merged**